manager/scheduler/nodeinfo.go

```diff
@@ -54,12 +105,49 @@ func (nodeInfo *NodeInfo) removeTask(t *api.Task) bool {
 	}
 
 	reservations := taskReservations(t.Spec)
-	nodeInfo.AvailableResources.MemoryBytes += reservations.MemoryBytes
-	nodeInfo.AvailableResources.NanoCPUs += reservations.NanoCPUs
+	resources := &nodeInfo.AvailableResources
+
+	resources.MemoryBytes += reservations.MemoryBytes
+	resources.NanoCPUs += reservations.NanoCPUs
+
+	taken := nodeInfo.ThirdPartyResourcesTaken
+
+	for k, v := range reservations.ThirdParty {
+		reclaimTPR(resources.ThirdParty[k], v, taken[t.ID][k])
+	}
+
+	delete(taken, t.ID)
 
 	return true
 }
 
+func setEnvFromTPR(t *api.Task, k string, nodeRes *api.ThirdPartyResource, taskRes *api.ThirdPartyResource, taken *api.Set) {
+	tr := taskRes.Resource.(*api.ThirdPartyResource_Integer).Integer.Val
+
+	switch nodeRes.Resource.(type) {
+	case *api.ThirdPartyResource_Integer:
+		setEnv(t, fmt.Sprintf("%s=%d", k, tr))
+	case *api.ThirdPartyResource_Set:
+		set := make([]string, len(taken.Val))
+
+		i := 0
+		for k := range taken.Val {
+			set[i] = k
+			i++
+		}
+
+		setEnv(t, fmt.Sprintf("%s=%s", k, strings.Join(set, ",")))
+	}
+}
+
+func setEnv(t *api.Task, env string) {
+	switch r := t.Spec.GetRuntime().(type) {
+	case *api.TaskSpec_Container:
+		r.Container.Env = append(r.Container.Env, env)
```

Comment:
Also, I just realized that if there's a leader switchover, the new leader will need to know which specific resource from a set was allocated to a task, so it doesn't assign that resource to a second task. Putting the allocations on the `Task` object would make this possible, but if the item from the set it only put into an environment variable, the information is essentially lost.